Fixed #31318 -- Allowed sqlmigrate to inspect squashed migrations.

David-Wobrock opened the pull request:
Hi!

https://code.djangoproject.com/ticket/31318
Currently when calling `sqlmigrate` on a migration that is being replaced by a squash migration (because the migration is in the `replaces` class attribute of the squad migration), then we get a KeyError.

I tackled the patch and I suggest a **commit-by-commit** review.

The approach I took is interfere as less as possible with the existing migration graph and keep its logic as it is. Currently, the migration graph seems to represent one migration hierarchy and database evolution schema. Intervening in the "forward state" logic in the graph would have affected lots of logic in the Django migration system.
Instead, I added a flag enabling loading the migration graph (by the migration loader obviously) without replacing the migrations by squashed migrations. This flag enables to keep entire logic as it is, but just have another representation of the migration graph that we will use instead and just for `sqlmigration`.

adamchainz commented on django/core/management/commands/sqlmigrate.py:
```diff
@@ -53,15 +53,20 @@ def handle(self, *args, **options):
+        # If migration was replaced, load the graph without replacements
+        # in order to show SQL plan
+        if target in executor.loader.graph.replaced_node_keys:
+            executor = MigrationExecutor(connection, replace_migrations=False)
```
Won't it be faster to just make the inital load not use replacement at all now?

David-Wobrock replied:
> Won't it be faster to just make the initial load not use replacement at all now?

It seems to also allow inspecting replacement/squash migrations when not loading them :+1: Seems logical since we still load them because they are in the disk_migrations, but we just don't remove the replaced migrations. Good call, thanks!

adamchainz commented on tests/migrations/test_commands.py:
```diff
@@ -695,6 +695,40 @@ def test_sqlmigrate_for_non_transactional_databases(self):
+        output = out.getvalue().lower()
+        self.assertNotEqual(output.find('-- create model author'), -1)
+        self.assertEqual(output.find('-- create model tribble'), -1)
+        self.assertNotEqual(output.find('-- create model book'), -1)
```
Please split into 4 tests, one for each thing being tested, and migrate those comments into the test names.

adamchainz commented:
I've renamed the PR into Django's archaic past tense format - please use this in future PR's :)

David-Wobrock commented:
Thanks for the review and renaming the PR @adamchainz :)
The last 2 commits take into account your comments

adamchainz commented on django/db/migrations/graph.py:
```diff
@@ -138,6 +139,7 @@ def remove_replaced_nodes(self, replacement, replaced):
         for replaced_key in replaced:
             self.nodes.pop(replaced_key, None)
             replaced_node = self.node_map.pop(replaced_key, None)
+            self.replaced_node_keys.add(replaced_key)
```
nitpick: This line woud be slightly more logical placed above the preceding one, grouping the manipulation of `replaced_key`, then `replaced_node`.

adamchainz commented on tests/migrations/test_commands.py:
```diff
@@ -695,6 +695,40 @@ def test_sqlmigrate_for_non_transactional_databases(self):
+        out = io.StringIO()
+        call_command("sqlmigrate", "migrations", "0001_initial", stdout=out)
+        output = out.getvalue().lower()
+        self.assertNotEqual(output.find('-- create model author'), -1)
```
n.b. just noticed these tests could also use `assertIn` / `assertNotIn` rather than `find()`. But it seems the tests in this file mix the two, so no worries.

David-Wobrock commented:
> Looks good to me.

Thanks! I took the nitpicks into account, see the last 2 commits again :)

felixxm commented on django/core/management/commands/sqlmigrate.py:
```diff
@@ -34,7 +34,8 @@ def handle(self, *args, **options):
         # Load up an executor to get all the migration data
-        executor = MigrationExecutor(connection)
+        # Do not replace migrations
+        executor = MigrationExecutor(connection, replace_migrations=False)
```
I think we can make small refactoring in advance. I would move `MigrationExecutor.collect_sql()` (which is used only by `sqlmigrate`) to the `MigrationLoader`. With this change we'll be able to use here `MigrationLoader` and simplify changes.

felixxm commented on django/core/management/commands/sqlmigrate.py:
```diff
@@ -53,15 +54,15 @@ def handle(self, *args, **options):
         # Make a plan that represents just the requested migrations and show SQL
         # for it
-        plan = [(executor.loader.graph.nodes[targets[0]], options['backwards'])]
+        plan = [(executor.loader.graph.nodes[target], options['backwards'])]
```
This small optimization is not related with fix. Merged in e12fea24f06f8911ddc2af1d6cbfb1adb529c1f2.

felixxm commented on tests/migrations/test_commands.py:
```diff
@@ -695,6 +695,40 @@ def test_sqlmigrate_for_non_transactional_databases(self):
         self.assertNotIn(connection.ops.end_transaction_sql().lower(), queries)
 
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_squashed"})
+    def test_sqlmigrate_ambigious_squashed_migration_name(self):
```
Two tests i.e. `test_sqlmigrate_ambigious_squashed_migration_name()` and `test_sqlmigrate_squashed_migration()` work without this patch. It's fine to increase coverage but we should move them to a separate commit.

felixxm commented on tests/migrations/test_commands.py:
```diff
@@ -695,6 +695,40 @@ def test_sqlmigrate_for_non_transactional_databases(self):
+        call_command("sqlmigrate", "migrations", "0002_second", stdout=out)
+        output = out.getvalue().lower()
+        self.assertIn('-- delete model tribble', output)
+        self.assertIn('-- create model book', output)
```
It looks that `test_sqlmigrate_replaced_second_migration()` and `test_sqlmigrate_replaced_migration()` are redundant. Please remove one of them.

felixxm commented on django/db/migrations/graph.py:
```diff
@@ -84,6 +84,7 @@ class MigrationGraph:
     def __init__(self):
         self.node_map = {}
         self.nodes = {}
+        self.replaced_node_keys = set()
```
`replaced_node_keys` is unused. Please revert changes in `django/db/migrations/graph.py`.

David-Wobrock commented:
Thanks for the review @felixxm !
I rebased my patch, made the changes and squashed the commits appropriately. I hope the changes suit now :)

felixxm commented:
@David-Wobrock Thanks :+1: I pushed minor edits, fixed commit messages, and added 271e108b29a1c280f20251b035fb8ca9a202b249 (`sqlmigrate` doesn't need `MigrationExecutor` anymore).

David-Wobrock commented:
> @David-Wobrock Thanks +1 I pushed minor edits, fixed commit messages, and added [271e108](https://github.com/django/django/commit/271e108b29a1c280f20251b035fb8ca9a202b249) (`sqlmigrate` doesn't need `MigrationExecutor` anymore).

Nice! Thanks :)
